### PR TITLE
Superseded Instances Remediation Policy

### PR DESCRIPTION
**Pull request by rshade**
### Description
This policy searches for an instance of a specific type and takes action on it. 

### Issues Resolved

[List any existing issues this PR resolves]

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

**Comment by rshade** [comment]
@avinashbhashyam-rs I thought the rs_cm. calls default to 1.6 if available? is there an option on those resources calls if not to use 1.6?

**Comment by avinashbhashyam-rs** [comment]
@rshade : No. rs.cm always uses API 1.5. and we don't have an option to specify it to use 1.6. You have to use right_api 1.6 as a http datasource.

**Comment by rshade** [comment]
@avinashbhashyam-rs would that break my excludes tags list?

**Comment by avinashbhashyam-rs** [comment]
@rshade : Yes. The `tags filter` is only available on resources. But the index call returns the tags set on the instance so you need to check for the exclude tags in java script.

File: cost/aws/superseded_instance/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+### What it does
+
+This Policy Template is used to automatically resize Amazon Resource Manager Instances (VMs) based on user-defined standards.  For example, the Policy allows for a Disallowed Instance Type and its corresponding Superseded Instance Type.  If any Instances are found using the Disallowed type, they will be included in the corresponing Policy Incident.  Upon approval, Cloud Workflow will resize the Instance to the Superseded type. 
+
+### Prerequesites
+- The following RightScale Credentials
+  - Cloud Management - `Actor`
+  - Cloud Management - `Observer`
+  - The `policy_designer`, `policy_manager` & `policy_publisher` roles
```
Comment (cdwilhelm) [inline cost/aws/superseded_instance/README.md]: update to be same format as Cloud Management

File: cost/aws/superseded_instance/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+## AWS Superseded Instance Types Remediation Policy
+
+### What it does
+
+This Policy Template is used to automatically resize Amazon Resource Manager Instances (VMs) based on user-defined standards.  For example, the Policy allows for a Disallowed Instance Type and its corresponding Superseded Instance Type.  If any Instances are found using the Disallowed type, they will be included in the corresponing Policy Incident.  Upon approval, Cloud Workflow will resize the Instance to the Superseded type.
```
Comment (cdwilhelm) [inline cost/aws/superseded_instance/README.md]: What is Amazon Resource Manager.  I think it should just say AWS .

Comment (cdwilhelm) [reply cost/aws/superseded_instance/README.md]: this sentence isn't clear. `For example, the Policy allows for a Disallowed Instance Type and its corresponding Superseded Instance Type.`   allows for what?

File: cost/aws/superseded_instance/README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+### Prerequesites
+- The following RightScale Credentials
+  - Cloud Management - `Actor`
+  - Cloud Management - `Observer`
+  - The `policy_designer`, `policy_manager` & `policy_publisher` roles
+
+### Functional Details
+
+- This policy supports a single old-to-new instance type mapping.  To apply this policy for multiple old-to-new instance type scenarios, the policy will need to be applied multiple times.  This will allow for flexibility in managing applied policies.
+- The Exclusion Tag parameter is a string value.  Supply the Tag Key only.  Tag Values are not analyzed and therefore are not need.  If the exclusion tag key is used on an Instance, that Instance is presumed to be exempt from this policy.
```
Comment (cdwilhelm) [inline cost/aws/superseded_instance/README.md]: Should be Exclude Tag

File: cost/aws/superseded_instance/aws_superseded_instance_remediation.pt [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+      "verb": "GET",
+      "host": rs_cm_host,
+      "path": cloud_href + "/instances",
+      "headers": {
+        "X-Api-Version": "1.6",
+        "X-Account": \""+rs_project_id+\""
+      },
+      "query_params": {
+        "view": "full",
+        "filter": "state=provisioned,operational" + filtered_tags
```
Comment (cdwilhelm) [inline cost/aws/superseded_instance/aws_superseded_instance_remediation.pt]: what about state=running?  isn't provisioned what raw instance report for stopped?

File: cost/aws/superseded_instance/aws_superseded_instance_remediation.pt [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+  end
+end
+
+define sys_log($subject, $detail) do
+  if $$debug
+    rs_cm.audit_entries.create(
+      notify: "None",
+      audit_entry: {
+        auditee_href: @@account,
+        summary: $subject,
```
Comment (cdwilhelm) [inline cost/aws/superseded_instance/aws_superseded_instance_remediation.pt]: Add the policy name to the subject

File: cost/aws/superseded_instance/aws_superseded_instance_remediation.pt [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+parameter "param_new_instance_type" do
+  label "Superseded Instance Type"
+  type "string"
+  description "New Instance Type"
+end
+
+parameter "param_exclude_tags" do
+  label "Exclude Tags"
+  description "list of excluded servers"
```
Comment (cdwilhelm) [inline cost/aws/superseded_instance/aws_superseded_instance_remediation.pt]: This param isn't the list of excluded server.  It's a list of tags to filter out instances from being checked.

File: cost/superseded_instance_remediation/superseded_instance_remediation.pt [hunk 7]
```diff
@@ -0,0 +1,218 @@
+name "Superseded Instance Remediation"
+rs_pt_ver 20180301
+type "policy"
+short_description "This Policy Template is used to automatically supersede instances based on user-defined standards. \n See the [README](https://github.com/rightscale/policy_templates/tree/master/cost/aws/aws_superseded_instance) and [docs.rightscale.com/policies](http://docs.rightscale.com/policies/) to learn more."
```
Comment (cdwilhelm) [inline cost/superseded_instance_remediation/superseded_instance_remediation.pt]: the link to the readme is wrong.

File: cost/superseded_instance_remediation/superseded_instance_remediation.pt [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+###############################################################################
+# Scripts
+###############################################################################
+
+script "js_instances_16", type: "javascript" do
+  result "request"
+  parameters "rs_cm_host","cloud_href","rs_project_id","param_exclude_tags"
+  code <<-EOS
+    var filtered_tags = ""
+    console.log(param_exclude_tags)
```
Comment (cdwilhelm) [inline cost/superseded_instance_remediation/superseded_instance_remediation.pt]: still needed?

File: cost/superseded_instance_remediation/README.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+- This policy supports a single old-to-new instance type mapping.  To apply this policy for multiple old-to-new instance type scenarios, the policy will need to be applied multiple times.  This will allow for flexibility in managing applied policies.
+- The Exclude Tag parameter is a string value.  Supply the Tag Key only.  Tag Values are not analyzed and therefore are not need.  If the exclusion tag key is used on an Instance, that Instance is presumed to be exempt from this policy.
+- Upon Approval, the Instances that appear in the Incident will be immediately stopped, resized, and started via Cloud Workflow. 
+
+#### Input Parameters
+
+This policy has the following input parameters required when launching the policy.
+
+- *Disallowed Instance Type* - a single Amazon instance type that should be resized
```
Comment (cdwilhelm) [inline cost/superseded_instance_remediation/README.md]: Remove Amazon, this is multi cloud, yes?

File: cost/superseded_instance_remediation/README.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+- *Superseded Instance Type* - the new Amazon instance type that instances should be resized to
```
Comment (cdwilhelm) [inline cost/superseded_instance_remediation/README.md]: Remove Amazon, this is multi cloud, yes?

File: cost/superseded_instance_remediation/README.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+  - Cloud Management - `Actor`
+  - Cloud Management - `Observer`
+  - Governance -  `policy_designer`
+  - Governance - `policy_manager`
+  - Governance - `policy_publisher`
+
+### Functional Details
+
+- This policy supports a single old-to-new instance type mapping.  To apply this policy for multiple old-to-new instance type scenarios, the policy will need to be applied multiple times.  This will allow for flexibility in managing applied policies.
+- The Exclude Tag parameter is a string value.  Supply the Tag Key only.  Tag Values are not analyzed and therefore are not need.  If the exclusion tag key is used on an Instance, that Instance is presumed to be exempt from this policy.
```
Comment (cdwilhelm) [inline cost/superseded_instance_remediation/README.md]: the parameter is a list, so can be more than one.

File: cost/superseded_instance_remediation/superseded_instance_remediation.pt [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+    check val(item,"locked")
+    escalate $report_and_resize_instances
+  end
+end
+
+###############################################################################
+# Cloud Workflow
+###############################################################################
+
+define resize_instances($data,$param_new_instance_type) do
```
Comment (cdwilhelm) [inline cost/superseded_instance_remediation/superseded_instance_remediation.pt]: does this need to check for 'locked' ?

File: cost/superseded_instance_remediation/README.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+### What it does
+
+This Policy Template is used to automatically resize instances based on user-defined standards.  For example, the Policy uses a Disallowed Instance Type and its corresponding Superseded Instance Type.  If any Instances are found using the Disallowed type, they will be included in the corresponing Policy Incident.  Upon approval, Cloud Workflow will resize the Instance to the Superseded type. 
+
+### Prerequesites
+- The following RightScale Credentials
+  - Cloud Management - `Actor`
+  - Cloud Management - `Observer`
+  - Governance -  `policy_designer`
```
Comment (srpomeroy) [inline cost/superseded_instance_remediation/README.md]: Let's start excluding the `policy_*` roles from docs. They are more or less inferred.

File: cost/superseded_instance_remediation/superseded_instance_remediation.pt [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+  EOS
+end
+
+###############################################################################
+# Escalations
+###############################################################################
+escalation "report_and_resize_instances" do
+  request_approval  do
+    label "Escalation approval"
+    description "Approve delete action"
```
Comment (srpomeroy) [inline cost/superseded_instance_remediation/superseded_instance_remediation.pt]: This description appears to be from another policy template.

File: cost/superseded_instance_remediation/superseded_instance_remediation.pt [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+        results.push(instance)
+      }
+    }
+  EOS
+end
+
+###############################################################################
+# Escalations
+###############################################################################
+escalation "report_and_resize_instances" do
```
Comment (srpomeroy) [inline cost/superseded_instance_remediation/superseded_instance_remediation.pt]: Needs to be updated for best practices. Also, recommend removing the word "report".
Suggest `esc_resize_instances`

File: cost/superseded_instance_remediation/superseded_instance_remediation.pt [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+
+resolution "report_and_resize_instances_resolution" do
+  email $param_email
+end
+
+###############################################################################
+# Policy
+###############################################################################
+
+policy "pol_superseded_instance" do
```
Comment (srpomeroy) [inline cost/superseded_instance_remediation/superseded_instance_remediation.pt]: Should be `policy_superseded_instance`.

File: cost/superseded_instance_remediation/superseded_instance_remediation.pt [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+      label "Reason for approval"
+      description "Explain why you are approving the action"
+    end
+  end
+
+  run "resize_instances", data, $param_new_instance_type
+  email $param_email
+end
+
+resolution "report_and_resize_instances_resolution" do
```
Comment (srpomeroy) [inline cost/superseded_instance_remediation/superseded_instance_remediation.pt]: Needs to be updated to match best practices. i.e. `res_resize_instances`
We should also remove the word report.

File: cost/superseded_instance_remediation/README.md [hunk 18]
```diff
@@ -0,0 +1,35 @@
+## Superseded Instance Types Remediation Policy
+
+### What it does
+
+This Policy Template is used to automatically resize instances based on user-defined standards.  For example, the Policy uses a Disallowed Instance Type and its corresponding Superseded Instance Type.  If any Instances are found using the Disallowed type, they will be included in the corresponing Policy Incident.  Upon approval, Cloud Workflow will resize the Instance to the Superseded type. 
+
+### Prerequesites
+- The following RightScale Credentials
```
Comment (cdwilhelm) [inline cost/superseded_instance_remediation/README.md]: these are not credentials they are permissions :)

File: cost/superseded_instance_remediation/README.md [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+  - Cloud Management - `Actor`
+  - Cloud Management - `Observer`
+
+### Functional Details
+
+- This policy supports a single old-to-new instance type mapping.  To apply this policy for multiple old-to-new instance type scenarios, the policy will need to be applied multiple times.  This will allow for flexibility in managing applied policies.
```
Comment (cdwilhelm) [inline cost/superseded_instance_remediation/README.md]: applying multiple policies may be confusing to a policy user.  i'm wondering if some examples should be giving.

File: cost/superseded_instance_remediation/superseded_instance_remediation.pt [hunk 20]
```diff
@@ -0,0 +1,304 @@
...
+severity "low"
+
+###############################################################################
+# Permissions
+###############################################################################
+
+permission "perm_instances" do
+  label "Permissions for instances"
+  actions "rs_cm.index", "rs_cm.show"
+  resources "rs_cm.instances"
```
Comment (cdwilhelm) [inline cost/superseded_instance_remediation/superseded_instance_remediation.pt]: do you need to add clouds?

File: cost/superseded_instance_remediation/superseded_instance_remediation.pt [hunk 21]
```diff
@@ -0,0 +1,304 @@
...
+###############################################################################
+# Scripts
+###############################################################################
+
+script "js_instances_16", type: "javascript" do
+  result "request"
+  parameters "rs_cm_host","cloud_href","rs_project_id","param_exclude_tags"
+  code <<-EOS
+    var filtered_tags = ""
+    console.log(param_exclude_tags[0])
```
Comment (cdwilhelm) [inline cost/superseded_instance_remediation/superseded_instance_remediation.pt]: should this be removed

File: cost/superseded_instance_remediation/superseded_instance_remediation.pt [hunk 22]
```diff
@@ -0,0 +1,304 @@
...
+###############################################################################
+# Policy
+###############################################################################
+
+policy "policy_superseded_instance" do
+  validate $ds_combined_instance_data do
+    summary_template "{{ rs_project_name }} (Account ID: {{ rs_project_id }}): {{ len data }} Instances will be resized from {{ parameters.param_old_instance_type }} to {{ parameters.param_new_instance_type }}"
+    detail_template <<-EOS
+# Superseded Instances
+| Resource UID | Name | Cloud |
```
Comment (cdwilhelm) [inline cost/superseded_instance_remediation/superseded_instance_remediation.pt]: include current and new instance type?  should the row indicate that an instance type can not be changed.

File: cost/superseded_instance_remediation/superseded_instance_remediation.pt [hunk 23]
```diff
@@ -0,0 +1,304 @@
...
+parameter "param_email" do
+  type "list"
+  label "Email addresses"
+  description "Email addresses of the recipients you wish to notify"
+end
+
+parameter "param_old_instance_type" do
+  label "Disallowed Instance Type"
+  type "string"
+  description "Old Instance Type"
```
Comment (cdwilhelm) [inline cost/superseded_instance_remediation/superseded_instance_remediation.pt]: Is it disallowed or old instance types.  I think they can be different.  Also include a better description of what the parameter is about.

**Pull request merged**